sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -181,6 +182,10 @@ case class DataSource(
       throw new AnalysisException(
         s"Unable to infer schema for $format. It must be specified manually.")
     }
+
+    SchemaUtils.checkSchemaColumnNameDuplication(
+      dataSchema, "the datasource", sparkSession.sessionState.conf.caseSensitiveAnalysis)
```

Comment:
Actually, it should be dataSchema + partitionSchema.

We also need to issue some meaningful error message. Users still can bypass the error by manually specify the schema.